Description: Fix kFreeBSD build
 Drop redundant sys/dirent.h include
Forwarded: https://github.com/htop-dev/htop/pull/386
Author: Graham Inggs <ginggs@debian.org>
Last-Update: 2020-12-11

--- a/freebsd/FreeBSDProcessList.c
+++ b/freebsd/FreeBSDProcessList.c
@@ -14,7 +14,6 @@
 #include <stdlib.h>
 #include <string.h>
 #include <sys/_iovec.h>
-#include <sys/dirent.h>
 #include <sys/errno.h>
 #include <sys/param.h> // needs to be included before <sys/jail.h> for MAXPATHLEN
 #include <sys/jail.h>
