From 898d429444623d1ce328053d3ea2348f1e737fe2 Mon Sep 17 00:00:00 2001
From: Benny Baumann <BenBE@geshi.org>
Date: Mon, 14 Apr 2025 19:10:55 +0200
Subject: [PATCH] Keep following when moving left or right

Fixes: #1674
---
 MainPanel.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/MainPanel.c b/MainPanel.c
index 7852c08f3..b665e0e26 100644
--- a/MainPanel.c
+++ b/MainPanel.c
@@ -117,6 +117,8 @@ static HandlerResult MainPanel_eventHandler(Panel* super, int ch) {
       result = HANDLED;
    } else if (0 < ch && ch < 255 && isdigit((unsigned char)ch)) {
       MainPanel_idSearch(this, ch);
+   } else if (ch == KEY_LEFT || ch == KEY_RIGHT) {
+      reaction |= HTOP_KEEP_FOLLOWING;
    } else {
       if (ch != ERR) {
          this->idSearch = 0;
